test/Reflection/capture_descriptors.sil

```diff
@@ -0,0 +1,207 @@
+  dealloc_stack %i : $*Int
+  return %c : $@callee_owned () -> ()
+}
+
+// CHECK:      - Capture types:
+// CHECK-NEXT: (struct Swift.Int)
+// CHECK-NEXT: (sil_box
+// CHECK-NEXT:   (struct Swift.String))
+// CHECK-NEXT:   - Metadata sources:
+
+
+// Generic caller and generic callee -- capture types are written in terms of
+// the caller's generic parameters, and metadata bindings are present for
+// structural sub-terms of the callee's substituted generic parameters that
+// contain the caller's generic parameters.
+
+sil @generic_callee3 : $@convention(thin) <T, U> (@in T) -> () {
+bb0(%t: $*T):
+  %12 = tuple ()
+  return %12 : $()
+}
+
+sil @generic_caller3 : $@convention(thin) <A, B, C> () -> @owned @callee_owned () -> () {
+bb0:
+  %f = function_ref @generic_callee3 : $@convention(thin) <T, U> (@in T) -> ()
+  %t = alloc_stack $Optional<A -> B>
+  %c = partial_apply %f<Optional<A -> B>, (B, C -> Int)>(%t) : $@convention(thin) <T, U> (@in T) -> ()
+  dealloc_stack %t : $*Optional<A -> B>
+  return %c : $@callee_owned () -> ()
+}
+
+// CHECK:      - Capture types:
+// CHECK-NEXT: (bound_generic_enum Swift.Optional
+// CHECK-NEXT:   (function
+// CHECK-NEXT:     (generic_type_parameter depth=0 index=0)
+// CHECK-NEXT:     (generic_type_parameter depth=0 index=1)))
+// CHECK-NEXT: - Metadata sources:
+// CHECK-NEXT: (bound_generic_enum Swift.Optional
+// CHECK-NEXT:   (function
+// CHECK-NEXT:     (generic_type_parameter depth=0 index=0)
+// CHECK-NEXT:     (generic_type_parameter depth=0 index=1)))
+// CHECK-NEXT: (closure_binding index=0)
+// CHECK-NEXT: (generic_type_parameter depth=0 index=1)
+// CHECK-NEXT: (closure_binding index=1)
+// CHECK-NEXT: (generic_type_parameter depth=0 index=2)
+// CHECK-NEXT: (closure_binding index=2)
+
+
+// Generic caller and generic callee -- and one of the type parameters is
+// fulfilled by a thick metatype value.
+
+sil @generic_callee4 : $@convention(thin) <T, U> (@thick Optional<T>.Type) -> () {
+bb0(%t: $@thick Optional<T>.Type):
+  %12 = tuple ()
+  return %12 : $()
+}
+
+sil @generic_caller4 : $@convention(thin) <A, B, C> () -> @owned @callee_owned () -> () {
+bb0:
+  %f = function_ref @generic_callee4 : $@convention(thin) <T, U> (@thick Optional<T>.Type) -> ()
+  %t = metatype $@thick Optional<B -> C>.Type
+  %c = partial_apply %f<B -> C, Int>(%t) : $@convention(thin) <T, U> (@thick Optional<T>.Type) -> ()
+  return %c : $@callee_owned () -> ()
+}
+
+// CHECK:      - Capture types:
+// CHECK-NEXT: (metatype was_abstract
+// CHECK-NEXT:  (bound_generic_enum Swift.Optional
+// CHECK-NEXT:    (function
+// CHECK-NEXT:      (generic_type_parameter depth=0 index=1)
+// CHECK-NEXT:      (generic_type_parameter depth=0 index=2))))
+// CHECK-NEXT: - Metadata sources:
+// CHECK-NEXT: (function
+// CHECK-NEXT:   (generic_type_parameter depth=0 index=1)
+// CHECK-NEXT:   (generic_type_parameter depth=0 index=2))
+// CHECK-NEXT: (generic_argument index=0
+// CHECK-NEXT:   (reference_capture index=0))
```

Comment:
Is this reference capture the `@thick Optional<T>.Type`? Seems like a weird name for it, if so. Is this basically "pointer capture argument" now, and you have to check the typeref of the capture to know whether it's a class instance or a type?